content/blog/the-rise-of-special-project-infra.md

```diff
@@ -0,0 +1,168 @@
+
+```nix
+{
+  networking.firewall.allowedTCPPorts = [ 80 443 ]
+    ++ [ config.services.owncast.rtmp-port ];
+
+  security.acme.defaults.email = "admin@example.com";
+
+  services.owncast.enable = true;
+
+  services.caddy = {
+    enable = true;
+    email = config.security.acme.defaults.email;
+    virtualHosts = {
+      "live.nixos.org".extraConfig = let
+        owncastWebService = "http://${config.services.owncast.listen}:${
+            toString config.services.owncast.port
+          }";
+      in ''
+        encode gzip
+        reverse_proxy ${owncastWebService}
+      '';
+    };
+  };
+}
+```
+
+Demonstrating the great UX of a well-written NixOS module, the Owncast configuration is completely out-of-the-box; using [all defaults](https://search.nixos.org/options?channel=unstable&from=0&size=50&sort=relevance&type=options&query=owncast), one line installs and activates the service, initializes all necessary state (including a SQLite database `owncast.db` that stores important service configuration such as administrative credentials) in `/var/lib/owncast`, and binds the web server to `127.0.0.1:8080`.
+
+The Caddy module is similarly loaded (with goodies). The only customization necessary here was to specify the domain we prepared (in the last step of the provisioning phase above) as a virtual host from which Caddy can forward client requests to the backend Owncast web server. For good measure, we also specified an email address for Caddy to use when setting up SSL on our behalf. (This is the email address to use for account creation and correspondence from the SSL certificate authority, which is [Let's Encrypt](https://letsencrypt.org/) by default.)
+
+The remaining configuration simply opens ports in the system firewall for HTTP, HTTPS, and RTMP. (Note that the RTMP port is also an Owncast module-specified default.)
+
+That’s the entirety of the NixOS configuration specific to our use case; all additional configuration (e.g., importing generated hardware configuration, enabling SSH and adding appropriate keys, defining Nix garbage collection rules, etc.) is simply boilerplate.
+
+### Phase IV: Deploy
+
+With the instance provisioned and the configuration prepared, it was about time to introduce the two, marking the beautiful beginning of a successful working relationship. This is generally a two-step process:
+
+1. Install NixOS on the instance. Note: a provider may make this very easy (by providing a template image), moderately easy (by allowing [a custom ISO](https://github.com/nix-community/nixos-generators) to be uploaded), or hit-or-miss (by providing an image for a non-NixOS distribution from which NixOS may be converted with [NixOS-Infect](https://github.com/elitak/nixos-infect/) or [`NIXOS_LUSTRATE`](https://nixos.org/manual/nixos/stable/#sec-installing-from-other-distro)).
+
+   Our provider not doing us too many favors but at least known to play nice with `nixos-infect`ed installations, we went with that.
```

Comment:
```suggestion
   Our provider was not doing us too many favors, but as it is at least known to play nice with `nixos-infect`ed installations, we went with that.
```

I suppose you meant that?

Reply:
While I don’t consider the original sentence ungrammatical – think `A doing B, C did D.` – I’m happy to restructure it for clarity.

Reply:
Yeah, it makes sense on second thought, but really hard to parse without punctuation. I'd prefer clarity over elegance every time.

Reply:
👍